frontend/catalyst/pennylane_extensions.py

```diff
@@ -1058,32 +1064,62 @@ class ForLoop(HybridOp):
 
     binder = for_p.bind
 
+    def __init__(self, *args, apply_reverse_transform: bool, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.apply_reverse_transform = apply_reverse_transform
+
     def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
         op = self
         inner_trace = op.regions[0].trace
         inner_tape = op.regions[0].quantum_tape
-        res_classical_tracers = op.regions[0].res_classical_tracers
+        expansion_strategy = self.expansion_strategy
 
         with EvaluationContext.frame_tracing_context(ctx, inner_trace):
             qreg_in = _input_type_to_tracers(inner_trace.new_arg, [AbstractQreg()])[0]
             qrp_out = trace_quantum_tape(inner_tape, device, qreg_in, ctx, inner_trace)
             qreg_out = qrp_out.actualize()
-            jaxpr, _, consts = ctx.frames[inner_trace].to_jaxpr2(res_classical_tracers + [qreg_out])
 
-        step = op.in_classical_tracers[2]
-        apply_reverse_transform = isinstance(step, int) and step < 0
-        qreg = qrp.actualize()
+            arg_expanded_tracers = expand_args(
+                self.regions[0].arg_classical_tracers + [qreg_in],
+                expansion_strategy=expansion_strategy,
+            )[0]
+
+            nimplicit = len(arg_expanded_tracers) - len(self.regions[0].arg_classical_tracers) - 1
+
+            res_expanded_tracers, _ = expand_results(
+                [],
+                arg_expanded_tracers,
+                self.regions[0].res_classical_tracers + [qreg_out],
+                expansion_strategy=expansion_strategy,
+                num_implicit_inputs=nimplicit,
+            )
+            jaxpr, _, consts = ctx.frames[inner_trace].to_jaxpr2(res_expanded_tracers)
+
+        in_expanded_tracers = [
+            *[trace.full_raise(c) for c in consts],
+            *expand_args(op.in_classical_tracers, expansion_strategy=expansion_strategy)[0],
+            qrp.actualize(),
+        ]
+
+        out_expanded_classical_tracers = expand_results(
```

Comment:
What does the expand results do?

Reply:
"Expanded" is a recurring term for list of arguments/results. `expanded_` in the name of the list means that the implicit shape arguments were already added to the beginning of this list. I am working on the description document where I have added a corresponding [section](https://github.com/PennyLaneAI/catalyst/blob/dynshape-quantum-primitives/FRONTEND.md#expandedcollapsed-arguments-or-results).